fdbserver/CommitProxyServer.actor.cpp

```diff
@@ -3828,6 +3877,21 @@ ACTOR Future<Void> commitProxyServerCore(CommitProxyInterface proxy,
 		when(TxnStateRequest request = waitNext(proxy.txnState.getFuture())) {
 			addActor.send(processTransactionStateRequestPart(&transactionStateResolveContext, request));
 		}
+		when(SetThrottledShardRequest request = waitNext(proxy.setThrottledShard.getFuture())) {
+			for (auto& shard : request.throttledShards) {
+				auto it = commitData.hotShards.begin();
+				for (; it != commitData.hotShards.end(); ++it) {
+					if (it->first == shard) {
+						it->second = request.expirationTime;
+						break;
+					}
```

Comment:
This may be over-simplified. what if the incoming request, i.e., new hot shard overlaps with existing hot shards?

A proper data structure is to use `KeyRangeMap`, which provides APIs to deal with such problems, as well as checking key/range overlap with existing ranges. See `struct ProxyCommitData` for examples.

Reply:
could the SS create two shards that overlap ? I don't see how that could happen.

Reply:
I think it's possible. Say initially the shard is `[a, e)`, but split into `[a, c), [c, d), [d, e)`, then got merged to `[d, f)`. Then `[d, f)` becomes hot, which overlaps with `[a, e)`.

Reply:
The PR can be modified to protect against this, but I think the PR's knobs are set to avoid situations where shard splitting and shard throttling are happening at the same time. i.e. we don't want throttling to prevent the DD from splitting when it should.

Reply:
Shards can be split and merged, but I don't think it changes behavior,

Current time is T. Say `[a, e]` is hot. Any mutations to it will be throttled for `HOT_SHARD_THROTTLING_EXPIRE_AFTER` seconds. This is set to 3, so mutations to `[a,e)` are throttled until  expiration at T+3 seconds.

Suppose `[d,f)` comes into being. It won't be declared "hot" because the RK only polls every `HOT_SHARD_MONITOR_FREQUENCY` seconds (5), so no new hot shards will be created until T+5.

For the sake of argument, say the polling was set to 2, and `[d,f)` became hot at time T+2. Any mutations to the original shard `[a,e)` would be throttled until T+3, and any mutations to `[d, f)` would be throttled until T+5. This is by design..